src/tools/w3dviewer/emitterlinegroupproppage.cpp

```diff
@@ -0,0 +1,277 @@
+{
+    CDialog::OnInitDialog();
+    m_blurTimeBar = static_cast<ColorBarClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_LINEBAR), "CLASSPOINTER"));
+    InitializeSpinButton(&m_blurTimeSpin, m_blurTimeKeyFrames.Rand, 0.0f, 10000.0f);
+    m_blurTimeBar->SetMinMax(0.0f, 1.0f);
+    m_blurTimeBar->Clear();
+    m_blurTimeBar->SetColor(0, 0.0f, 0.0f, 0.0f, 0.0f, 3);
+    m_blurTimeBar->SetGradientValue(0, (m_blurTimeKeyFrames.Start - m_minBlurTime) / (m_maxBlurTime - m_minBlurTime));
+
+    for (unsigned int i = 0; i < m_blurTimeKeyFrames.NumKeyFrames; i++) {
+        m_blurTimeBar->SetColor(i + 1, m_blurTimeKeyFrames.KeyTimes[i] / m_lifetime, 0.0f, 0.0f, 0.0f, 3);
+        m_blurTimeBar->SetGradientValue(
+            i + 1, (m_blurTimeKeyFrames.Values[i] - m_minBlurTime) / (m_maxBlurTime - m_minBlurTime));
+    }
+
+    return TRUE;
+}
+
+void EmitterLineGroupPropPageClass::UpdateLifetime(float lifetime)
+{
+    if (m_lifetime != lifetime) {
+        float val = lifetime / m_lifetime;
+
+        for (unsigned int i = 0; i < m_blurTimeKeyFrames.NumKeyFrames; i++) {
+            m_blurTimeKeyFrames.KeyTimes[i] *= val;
+        }
+
+        m_instanceList->Set_Blur_Time_Keyframes(m_blurTimeKeyFrames);
+        m_lifetime = lifetime;
+    }
+}
+
+void EmitterLineGroupPropPageClass::Initialize()
+{
+    if (m_blurTimeKeyFrames.KeyTimes != nullptr) {
+        delete[] m_blurTimeKeyFrames.KeyTimes;
+        m_blurTimeKeyFrames.KeyTimes = nullptr;
+    }
+
+    if (m_blurTimeKeyFrames.Values != nullptr) {
+        delete[] m_blurTimeKeyFrames.Values;
+        m_blurTimeKeyFrames.Values = nullptr;
+    }
+
+    if (m_instanceList != nullptr) {
+        m_lifetime = m_instanceList->Get_Lifetime();
+        m_instanceList->Get_Blur_Time_Keyframes(m_blurTimeKeyFrames);
+        int max = m_blurTimeKeyFrames.Start;
+
+        if (max <= 1.0f) {
+            max = 1.0f;
+        }
+
+        m_maxBlurTime = max;
+        int min = m_blurTimeKeyFrames.Start;
+
+        if (min >= 0.0f) {
+            min = 0.0f;
+        }
+
+        m_minBlurTime = min;
+
+        for (unsigned int i = 0; i < m_blurTimeKeyFrames.NumKeyFrames; i++) {
+            if (m_blurTimeKeyFrames.Values[i] > m_maxBlurTime) {
+                m_maxBlurTime = m_blurTimeKeyFrames.Values[i];
+            }
+
+            if (m_blurTimeKeyFrames.Values[i] < m_minBlurTime) {
+                m_minBlurTime = m_blurTimeKeyFrames.Values[i];
+            }
+        }
+    }
+}
+
+void EmitterLineGroupPropPageClass::UpdateBlurTime()
+{
+    float time = 0.0f;
+    float red = 0.0f;
+    float green = 0.0f;
+    float blue = 0.0f;
+    float value = m_blurTimeBar->GetGradientValue(0);
+    m_blurTimeKeyFrames.Start = (m_maxBlurTime - m_minBlurTime) * value + m_minBlurTime;
+
+    if (m_blurTimeKeyFrames.KeyTimes != nullptr) {
+        delete[] m_blurTimeKeyFrames.KeyTimes;
+    }
+
+    if (m_blurTimeKeyFrames.Values != nullptr) {
+        delete[] m_blurTimeKeyFrames.Values;

```

Comment:
If `count` below is <= 1, then this becomes dangling pointer.

`m_blurTimeKeyFrames.Values = nullptr` missing.

Reply:
fixed.